rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+2. Delete the writer's checkpoint, if any.
+3. Update the manifest by setting the `destroyed_at_s` field to the current time.
+
+A hard delete will proceed as follows. It is up to the user to ensure that there is no active writer at the time of deletion:
+1. Check that there are no active checkpoints. If there are, return `InvalidDeletion`
+2. Delete all objects under `path`.
+
+### Readers
+
+Readers are created by calling `DbReader::open`. `open` takes an optional checkpoint. If a checkpoint is provided, `DbReader` uses the provided checkpoint and does not try to refresh it. If no checkpoint is provided, `DbReader` establishes it’s own checkpoint against the latest manifest and periodically re-establishes this checkpoint at the interval specified in `manifest_poll_interval`. The checkpoints created by `DbReader` use the lifetime specified in the `checkpoint_lifetime` option.
+
+#### Establishing the Checkpoint
+
+To establish the checkpoint, `DbReader` will:
+1. Read the latest manifest version V.
+2. Create a new checkpoint with id set to a new v4 uuid, `manifest_id` set to the next manifest version (V + 1), `checkpoint_expire_time_s` set to the current time plus `checkpoint_lifetime`, and `metadata` set to None.
+3. Update the manifest’s `wal_id_last_seen` to the latest WAL id found in the wal directory
+4. Write the manifest. If this fails with a CAS error, go back to 1.
+
+Then, the reader will restore data from the WAL. The reader will just replay the WAL data into an in-memory table. The size of the table will be bounded by `max_memtable_bytes`. When the table fills up, it’ll be added to a list and a new table will be initialized. The total memory used should be limited by backpressure applied by the writer. This could still be a lot of memory (e.g. up to 256MB in the default case). In the future we could have the reader read directly from the WAL SSTs, but I’m leaving that out of this proposal.
+
+##### Checkpoint Maintenance
+
+If the reader is opened without a checkpoint, it will periodically re-establish a checkpoint by re-running the initialization process described above, with the following modifications:
+- It will delete the checkpoint that it had previously created (when we support range scans, we’ll need to reference count the checkpoint).
+- It will purge any in-memory tables with content from WAL SSTs lower than the new `wal_id_last_compacted`.
+
+#### Garbage Collector
+
+The GC task will be modified to handle soft deletes and manage checkpoints/clones. The GC tasks's algorithm will now look like the following:
+1. Read the current manifest. If no manifest is found, exit.
+2. If `destroyed_at_s` is set, and the current time is after `destroyed_at_s` + `db_delete_grace`, and there are no active checkpoints, then:
```

Comment:
is `db_delete_grace` a new db property? Can this be added to the public api section?

Reply:
Yep, I'll add that.